### PR TITLE
Fix chatinterface multimodal bug

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Closes: #9107

Consolidates some logic across stream and non-stream code-paths to make the code more maintainable.
Also adds more e2e tests for multimodal cases

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/chat_interface.py [hunk 1]
```diff
@@ -579,14 +579,12 @@ def response_as_dict(self, response: MessageDict | Message | str) -> MessageDict
...
         self,
         message: str | MultimodalData,
         history_with_input: TupleFormat | list[MessageDict],
-        request: Request,
-        *args,
-    ) -> tuple[TupleFormat, TupleFormat] | tuple[list[MessageDict], list[MessageDict]]:
-        if self.multimodal and isinstance(message, MultimodalData):
+    ) -> tuple[str | dict, TupleFormat | list[MessageDict]]:
+        if isinstance(message, MultimodalData):
             remove_input = (
```
Comment (abidlabs) [inline gradio/chat_interface.py:588]: nit: could refactor to `remove_input = len(message.files) + int(message.text is not None)`

File: gradio/chat_interface.py [hunk 2]
```diff
@@ -579,14 +579,12 @@ def response_as_dict(self, response: MessageDict | Message | str) -> MessageDict
             new_response = response
         return cast(MessageDict, new_response)
 
-    async def _submit_fn(
+    def _format_input(
```
Comment (abidlabs) [inline gradio/chat_interface.py:582]: nit: perhaps we should this method a more descriptive name? it took me a while to understand what its used for. Maybe `process_message_and_trim_history`?

File: gradio/chat_interface.py [hunk 3]
```diff
@@ -597,7 +595,26 @@ async def _submit_fn(
         else:
             history = history_with_input[:-1]
             message_serialized = message
+        return message_serialized, history
 
+    def _append_history(self, history, message, first_response=True):
+        if self.type == "messages":
+            message = self.response_as_dict(message)
+        if self.type == "tuples":
+            history[-1][1] = message  # type: ignore
```
Comment (abidlabs) [inline gradio/chat_interface.py:604]: Wouldn't it be cleaner just to have 
```
        if self.type == "tuples":
            history[-1][1] = message  # type: ignore
       else:
            message = self.response_as_dict(message)
            if first_response: 
               ..
```

**Review by abidlabs — APPROVED** [review]
Looks good @freddyaboulton! I can confirm this fixes the issue, and thanks for cleaning up the code. 

That being said, I think we have a lot of message formats for Chatbot now and I'm quite lost myself about how to use everything. For example, how would I modify the chat function [in this issue](https://github.com/gradio-app/gradio/issues/9107) if I wanted to include a file being streamed. Is that possible?

**Comment by freddyaboulton** [comment]
It doesn't have to be just text, it can be any valid Chatbot response, component instances included. 

I think the issue is that chatinterface only allows the bot to send one response at a time. For sending files, it's best to return a component (e.g. gr.Image), but you can't send both a component and text in the same message. It's possible with a regular Blocks app though. So maybe we can make ChatInterface return more than one message at a time?

**Comment by abidlabs** [comment]
Ah right, I was confused, I thought you might be able to return `{"text": ..., "files": []}` as well if the chatbot is multimodal. I just need to have a clearer mental picture / we might need to improve documentation here. 

> I think the issue is that chatinterface only allows the bot to send one response at a time. For sending files, it's best to return a component (e.g. gr.Image), but you can't send both a component and text in the same message. It's possible with a regular Blocks app though. So maybe we can make ChatInterface return more than one message at a time?

This _has_ been requested, although I'm not sure if its common enough to warrant adding yet another acceptable return type. Anyways, for a separate PR. This PR lgtm

**Comment by freddyaboulton** [comment]
Agreed re documentation revamp.

**Pull request merged**